File: Arrays_easy/rotate_array.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size()-1;
        while(k-->0){
            int temp = nums[n];
            for(int i = n-1; i>=0; i--)
                nums[i] = nums[i-1];
            nums[0] = temp;
        }
    }
};

int main()
{
    Solution obj;
    vector<int> nums = {1, 2, 3, 4, 5, 6, 7};
    int k = 3;
    for(int i=0;i<nums.size();i++)
        cout<<nums[i]<<" ";
    cout<<"\n";
    obj.rotate(nums, k);
    for(int i=0;i<nums.size();i++)
        cout<<nums[i]<<" ";
    return 0;
}

File: Arrays_easy/rotate_array.py
# class Solution:
#     def rotate(self, nums: [int], k: int) -> None:
#         """
#         Do not return anything, modify nums in-place instead.
#         """
#         for _ in range(k):
#         	nums.insert(0, nums.pop(len(nums)-1))

# nums = [1,2,3,4,5,6,7]
# k = 3
# print("nums before rotating: {}".format(nums))
# obj = Solution()
# obj.rotate(nums, k)
# print("nums after rotating: {}".format(nums))

# using cyclic replacement, from leetcode
# class Solution:
#     def rotate(self, nums: [int], k: int) -> None:
#         """
#         Do not return anything, modify nums in-place instead.
#         """
#         n = len(nums)
#         k %= n
#         start = count = 0
#         while count < n:
#         	current = start
#         	prev = nums[current]
#         	while True:
#         		next_index = (current + k) % n
#         		nums[next_index], prev = prev, nums[next_index]
#         		current = next_index
#         		count += 1
#         		if current == start:
#         			break
#         	start += 1
				
# nums = [1,2,3,4,5,6,7]
# k = 3
# print("nums before rotating: {}".format(nums))
# obj = Solution()
# obj.rotate(nums, k)
# print("nums after rotating: {}".format(nums))

# reversal of array
class Solution:
	def reverse_arr(self, nums, start, end):
		while start < end:
			nums[start], nums[end] = nums[end], nums[start]
			start += 1
			end -= 1

	def rotate(self, nums: [int], k: int) -> None:
		"""
		Do not return anything, modify nums in-place instead.
		"""
		n = len(nums)
		k %= n
		self.reverse_arr(nums, 0, n-1)
		self.reverse_arr(nums, 0, k-1)
		self.reverse_arr(nums, k, n-1)

nums = [1,2,3,4,5,6,7]
k = 3
print("nums before rotating: {}".format(nums))
obj = Solution()
obj.rotate(nums, k)
print("nums after rotating: {}".format(nums))
